Redesign Global Fragment following Material Design Guidelines

ruif3r opened the pull request:
# Description

Redesign whole global fragment, using material cards and a pie chart, now its more easy to track the global progression of COVID. Other minor bugfixes

# ChangeType

Design changes and bugfixes

# Steps to Test

Open Global Fragment
Open Settings Activity from actionBar menu

ruif3r committed:
Redesign whole global fragment, using material cards and a pie chart, now its more easy to track the global progression of COVID. Other minor bugfixes

ruif3r committed:
Fixed wrong layout look on landscape on fragment_global.xml

ruif3r committed:
Merge branch 'master' into ruif3r/improve/redesign-global-fragment

ruif3r committed:
Merge branch 'master' into ruif3r/improve/redesign-global-fragment

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/res/values-night/styles.xml
#	app/src/main/res/values/styles.xml

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -94,6 +106,37 @@ class GlobalFragment : Fragment() {
+        colors.add(ContextCompat.getColor(requireContext(), R.color.recoveredColor))
+        colors.add(ContextCompat.getColor(requireContext(), R.color.deathsColor))
+        return colors
+    }
```
This is not that important I will leave the comment just so you take this into account. All this methods control the chart config, maybe they could be moved to a class in charge of configuring all the pie chart specific code and configurations, this would clean up the fragment and all the responsability would be moved to this new class.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -114,5 +157,11 @@ class GlobalFragment : Fragment() {
+fun formatLargeNumbers(value: Int): String {
+    val df = DecimalFormat("###,###,###")
+    return df.format(value)
+}
```
this should not be in the fragment. It is public and it doesn't have anything to do with the fragment, It should be moved to an utility formatter file. This way you could also reuse it somewhere else in your project.

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+
+        <com.github.mikephil.charting.charts.PieChart
+            android:id="@+id/circular_chart"
+            android:layout_width="150dp"
```
dimens

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+        <com.github.mikephil.charting.charts.PieChart
+            android:id="@+id/circular_chart"
+            android:layout_width="150dp"
+            android:layout_height="150dp"
```
dimens

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+            android:id="@+id/circular_chart"
+            android:layout_width="150dp"
+            android:layout_height="150dp"
+            android:layout_marginStart="24dp"
```
dimens

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+            android:id="@+id/last_update_date"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginEnd="24dp"
```
dimens

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+            android:id="@+id/last_update"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="24dp"
```
dimens

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+            android:layout_height="wrap_content"
+            android:layout_marginTop="16dp"
+            android:text="@string/global"
+            android:textSize="20sp"
```
dimens

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:fitsSystemWindows="true"
+        tools:context=".ui.global.GlobalFragment"
```
do you need this context?

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:fillViewport="true"
+    android:orientation="vertical">
```
is this orientation needed?.

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+
+
+        <com.google.android.material.card.MaterialCardView
+            android:id="@+id/materialCardView"
```
your Id's should follow the same convention, on the previous views you have `x_y_view_type` and now you have `xYyZz`, follow your own patterns.

jinmyjbv commented on app/src/main/res/layout/fragment_global.xml:
```diff
@@ -0,0 +1,225 @@
+        </com.google.android.material.card.MaterialCardView>
+
+        <com.google.android.material.card.MaterialCardView
+            android:id="@+id/materialCardView2"
```
in this view you have the same view structure 3 times, 

`MeterialCardView`
 -> `ConstraintLayout`
  -> `TextView`
  -> `TextView`
  -> `TextView`

as a good experiment, you could create your own custom view which contains this structure and that it receives the data that it needs to build and update itself.

ruif3r committed:
Lot of refactor changes following PR suggestions

-Moved fragment_global.xml piechart configuration  code from GlobalFragment.kt into PieChartBuilder.kt

-Hardcoded dimensions into dimens.xml

-Created a compound view for fragment_global.xml cardViews named CovidCustomCard.kt

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+const val DEFAULT_NUMBER_TEXT = "0"
+const val DEFAULT_PERCENTAGE_TEXT = "0%"
+const val CARD_ELEVATION = 10F
+const val CARD_RADIUS = 16F
```
why public?

jinmyjbv replied:
is this the best place to declare elevation and radius for the card?

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+        radius = CARD_RADIUS
+    }
+
+    override fun onFinishInflate() {
```
I think you don't need to wait for inflation finish, you can just create a method `initView()` and call it at the end of the `init {}` block, inflation is done almost immediately.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+
+            } finally {
+                recycle()
+            }
```
this process of getting variables from the styleable attributes could be abstracted to a different method.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+package com.example.ncov19traking.ui
+
+import android.content.Context
+import android.graphics.Color
```
this is not being used.

ruif3r replied:
I forgot configuring automatic optimize import

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+import com.example.ncov19traking.R
+import com.google.android.material.card.MaterialCardView
+
+const val DEFAULT_TITLE_TEXT = "Title"
```
What happens when you need internationalization, what problem will you have here? fix it.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+    private lateinit var percentageTextView : TextView
+
+    init {
+        setupCard()
```
so you are setting up view configuration before the inflation here? but at the same time you are waiting for inflation for other things on the `onFinishInflate` which one is it?.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,86 @@
+    private var percentageTextColor = resources.getColor(R.color.textColor)
+    private lateinit var titleTextView : TextView
+    private lateinit var numbersTextView : TextView
+    private lateinit var percentageTextView : TextView
```
if you put your `findByIds` at the end of the init block you can stop using late init var here that are not bad but are expensive, and in this custom view you are not gaining anything by using them.

ruif3r replied:
I used lateinit to have the view available for the setters. May I ask you for another way to do this? so I can put it in practice

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -64,28 +68,36 @@ class GlobalFragment : Fragment() {
+                pieChart.getChartData(it)
+                pieChart.setVisibility(true)
                 progressBar.isVisible = false
             }
```
this whole block could be separated in different methods to make it more readable, I see three different things being done, setting percentages, configuring piechart, and setting case numbers, try to be careful with this things too make your code more organized.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -106,13 +118,11 @@ class GlobalFragment : Fragment() {
+        val covidDeathsCard: CovidCustomCard = view.findViewById(R.id.covid_deaths_cases_card)
         val lastUpdateLong: TextView = view.findViewById(R.id.last_update_date)
         val progressBar: ProgressBar = view.findViewById(R.id.progressBar)
+        val pieChart = PieChartBuilder(view.findViewById(R.id.circular_chart))
```
mmmm I don't like the naming of this, a `PieChartBuilder` seems like a class who creates piecharts, but this class is not actually building anything is doing configurations in an specific view, you can ignore this method but is something to think about maybe next time we talk I can explain to you what I'm thinking.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/utils/NumberFormatter.kt:
```diff
@@ -0,0 +1,10 @@
+fun formatLargeNumbers(value: Int): String {
+    val df = DecimalFormat("###,###,###")
+    return df.format(value)
+}
```
so the intention was to create a utility class named `NumberFormatter` but here you have a class and a method that are not related to each other.

ruif3r replied:
Can you explain me a bit more? shall I use companion objects?

ruif3r replied:
fixed

ruif3r committed:
Changes made following PR suggestion

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,98 @@
+    private var percentageTextView: TextView? = null
+
+    init {
+        receivingAttributes(context, attrs, defStyleAttr)
```
often method naming is done in infinitive form like `receiveAttributes`, `receive`, `parse`, `get`, not `receiving`, `parsing` or `getting`

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,98 @@
+                        percentageTextColor
+                    )
+
+                } finally {
```
do you need this finally block here? can this throw an exception?

ruif3r replied:
it throws a RuntimeException if the typedArray has already been recycled somewhere.

Take a look

https://developer.android.com/training/custom-views/create-view#applyattr

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,98 @@
+    private fun inflateView(context: Context) {
+        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+        inflater.inflate(R.layout.covid_custom_card, this@CovidCustomCard)
+    }
```
methods ordering, private last.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,98 @@
+        setCardBackgroundColor(ContextCompat.getColor(context, R.color.backgroundColor))
+        cardElevation = resources.getDimension(R.dimen.custom_covid_card_elevation)
+        radius = resources.getDimension(R.dimen.custom_covid_card_radius)
+        isClickable = true
```
can't you define this clickable directly on the xml? since it seems it never changes.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/CovidCustomCard.kt:
```diff
@@ -0,0 +1,98 @@
+    private fun initializeView() {
+        setCardBackgroundColor(ContextCompat.getColor(context, R.color.backgroundColor))
+        cardElevation = resources.getDimension(R.dimen.custom_covid_card_elevation)
+        radius = resources.getDimension(R.dimen.custom_covid_card_radius)
```
nice 👍

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -61,31 +67,18 @@ class GlobalFragment : Fragment() {
-                recovered.text = it.recovered.toString()
-                deaths.text = it.deaths.toString()
+                setCovidDataToCards(it)
+                receivingPercentageTextToCards(it)
```
same infinitive comment I left before

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -94,9 +87,54 @@ class GlobalFragment : Fragment() {
+        )
+    }
+
+    private fun GlobalFragmentViewHolder.setCovidDataToCards(
```
this name could improve, you are only calling `setCaseNumbers` so I thing it should be better to reflect that in the method name, `data` is too broad It implies you are setting different kind of models inside this method.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/ui/global/GlobalFragment.kt:
```diff
@@ -106,13 +144,12 @@ class GlobalFragment : Fragment() {
-        val progressBar: ProgressBar = view.findViewById(R.id.progressBar)
+        val progressBar: ProgressBar = view.findViewById(R.id.progress_bar)
+        val pieChart = view.findViewById<PieChart>(R.id.circular_chart)
+            .apply { PieChartConfigHelper.setupPieChart(this) }
```
It's better but let's better not do any kind of initialization here, you are initializing on other places of the fragment as well, why not do it outside in the on view created or where you are doing the other stuff, this seems really out of place here.

jinmyjbv commented on app/src/main/java/com/example/ncov19traking/PieChartConfigHelper.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    fun setAndConfigChartData(nCoVInfo: NCoVInfo, pieChart: PieChart, vararg colorList: Int) {
+        val pieEntry = ArrayList<PieEntry>()
```
use apply.

ruif3r committed:
Changes made following PR suggestion

ruif3r committed:
Merge branch 'master' into ruif3r/improve/redesign-global-fragment

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	build.gradle
#	gradle/wrapper/gradle-wrapper.properties
